src/model_selection/kFold.ts

```diff
@@ -0,0 +1,161 @@
+   * splitting. Defaults to `false`.
+   */
+  shuffle?: boolean
+  /**
+   * Random seed to be used for shuffling. Ignored if `nSplits = false`.
+   */
+  randomState?: number
+}
+
+/**
+ * K-Fold cross-validator
+ *
+ * Generates train and test indices to split data in train/test subsets.
+ * To generate these subsets, the dataset is split into k (about) evenly
+ * sized chunks of consecutive elements. Each split takes another chunk
+ * as test data and the remaining chunks are combined to be the training
+ * data.
+ *
+ * Optionally, the indices can be shuffled before splitting it into chunks
+ * (disabled by default).
+ *
+ * @example
+ * ```js
+ * import { KFold } from 'scikitjs'
+ *
+ * const kf = new KFold({ nSplits: 3 })
+ *
+ * const X = tf.range(0, 7).reshape([7, 1]) as Tensor2D
+ *
+ * console.log( 'nSplits:', kf.getNumSplits(X) )
+ *
+ * for (const { trainIndex, testIndex } of kf.split(X) )
+ * {
+ *   try {
+ *     console.log( 'train:', trainIndex.toString() )
+ *     console.log( 'test:',   testIndex.toString() )
+ *   }
+ *   finally {
+ *     trainIndex.dispose()
+ *      testIndex.dispose()
+ *   }
+ * }
+ * ```
+ */
+export class KFold implements CrossValidator {
+  nSplits: number
+  shuffle: boolean
+  randomState?: number
+
+  constructor({
+    nSplits = 5,
+    shuffle = false,
+    randomState
+  }: KFoldParams = {}) {
+    nSplits = Number(nSplits)
+    assert(
+      nSplits % 1 === 0 && nSplits > 1,
+      'new KFold({nSplits}): nSplits must be an int greater than 1.'
+    )
+    this.nSplits = nSplits
+    this.shuffle = !!shuffle
+    this.randomState = randomState
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public getNumSplits(X?: Tensor2D, y?: Tensor1D, groups?: Tensor1D): number {
+    return this.nSplits
+  }
+
+  public *split(
+    X: Tensor2D,
+    y?: Tensor1D,
+    groups?: Tensor1D
+  ): IterableIterator<{ trainIndex: Tensor1D; testIndex: Tensor1D }> {
+    const { nSplits, shuffle, randomState } = this
+
+    const nSamples = X.shape[0]
+
+    assert(
+      nSplits <= nSamples,
+      'KFold({nSplits})::split(X): nSplits must not be greater than X.shape[0].'
+    )
+
+    if (null != y) {
+      assert(
+        nSamples === y.shape[0],
+        'KFold::split(X,y): X.shape[0] must equal y.shape[0].'
+      )
+    }
+
+    if (null != groups) {
+      assert(
+        nSamples === groups.shape[0],
+        'KFold::split(X,y,groups): X.shape[0] must equal groups.shape[0].'
+      )
+    }
+
+    const range = new Int32Array(nSamples)
+    for (let i = 0; i < range.length; i++) {
+      range[i] = i
+    }
+    if (shuffle) {
+      const rng = alea(randomState?.toString())
+      randUtils.shuffle(rng)(range)
+    }
+
+    const chunkBase = Math.floor(nSamples / nSplits)
+    let remainder = nSamples % nSplits
+
+    for (let offset = 0; offset < nSamples; ) {
+      const chunk = chunkBase + Number(remainder-- > 0)
```

Comment:
Not sure what this is. Valid JS syntax?

Reply:
Yeah it is :smile: `Number(true) === 1` and `Number(false) === 0`. Basically the first `remainder` splits/chunks are one element larger than `chunkBase` to match sample size which are not evenly divisible by `nSplits`. It's equivalent to:
```js
const chunk = remainder-- > 0 ? chunkBase + 1 : chunkBase
```
Would that be more readable? I could change it.

Reply:
Same here. Ternary is more readable 😀